File: python-tests/reporting/summary.py
import json
from datetime import datetime
import platform
from typing import Dict, Any, List
from etypes import JsonLogicEngine

class TestSummary:
    def __init__(self):
        self.current_run: Dict[str, Dict[str, Dict]] = {}
        self.current_totals: Dict[str, Dict] = {}
        self.results_by_suite: Dict[str, Dict[str, Dict]] = {}
        self.totals: Dict[str, Dict] = {}

    def add_result(self, suite_name: str, engine: JsonLogicEngine, passed: int, total: int):
        # Initialize current run results if not exists
        if suite_name not in self.current_run:
            self.current_run[suite_name] = {}

        # Add suite result for current run
        engine_key = engine.value if isinstance(engine, JsonLogicEngine) else str(engine)
        stats = {
            "passed": passed,
            "total": total,
        }
        
        self.current_run[suite_name][engine_key] = stats

        # Update current totals
        if engine_key not in self.current_totals:
            self.current_totals[engine_key] = {"passed": 0, "total": 0, "success_rate": 0}
        
        self.current_totals[engine_key]["passed"] += passed
        self.current_totals[engine_key]["total"] += total

    def load_existing_results(self, data: Dict[str, Any]):
        self.results_by_suite = data.get("test_suites", {})
        self.totals = data.get("totals", {})

    def save_json(self, filename: str):
        # Merge current run with existing results
        for suite_name, suite_results in self.current_run.items():
            if suite_name not in self.results_by_suite:
                self.results_by_suite[suite_name] = {}
            self.results_by_suite[suite_name].update(suite_results)

        # Merge current totals with existing totals
        self.totals.update(self.current_totals)

        result = {
            "test_suites": self.results_by_suite,
            "totals": self.totals,
        }

        with open(filename, 'w') as f:
            json.dump(result, f, indent=2)

    def print_summary(self):
        print("\n" + "=" * 30 + " Test Results Summary " + "=" * 30)
        
        # Print header
        print(f"{'Test Suite':<30}", end="")
        for engine in JsonLogicEngine:
            print(f"{engine.value:>15}", end="")
        print("\n" + "-" * 80)

        # Print results for current run
        for suite_name in sorted(self.current_run.keys()):
            print(f"{suite_name:<30}", end="")
            for engine in JsonLogicEngine:
                stats = self.current_run[suite_name].get(engine.value)
                if stats:
                    print(f"{stats['passed']:>7}/{stats['total']:<7}", end="")
                else:
                    print(f"{'N/A':>15}", end="")
            print()

        # Print totals for current run
        print("-" * 80)
        print(f"{'TOTAL':<30}", end="")
        for engine in JsonLogicEngine:
            if engine.value in self.current_totals:
                stats = self.current_totals[engine.value]
                print(f"{stats['passed']:>7}/{stats['total']:<7}", end="")
            else:
                print(f"{'N/A':>15}", end="")
        print()

File: go-tests/reporting/summary.go
package reporting

import (
	"encoding/json"
	"fmt"
	"os"
	"runtime"
	"sort"
	"strings"
	"time"

	"jsonlogic-tests/types"
)

type EngineStats struct {
	Passed int `json:"passed"`
	Total  int `json:"total"`
}

func NewEngineStats(passed, total int) *EngineStats {
	return &EngineStats{
		Passed: passed,
		Total:  total,
	}
}

type SuiteResults map[types.JsonLogicEngine]*EngineStats

type TestSummary struct {
	resultsBySuite map[string]SuiteResults
	totals         map[types.JsonLogicEngine]*EngineStats
}

func NewTestSummary() *TestSummary {
	return &TestSummary{
		resultsBySuite: make(map[string]SuiteResults),
		totals:         make(map[types.JsonLogicEngine]*EngineStats),
	}
}

func (s *TestSummary) AddResult(suiteName string, engine types.JsonLogicEngine, passed, total int) {
	// Initialize suite results if not exists
	if _, ok := s.resultsBySuite[suiteName]; !ok {
		s.resultsBySuite[suiteName] = make(SuiteResults)
	}

	// Add suite result
	s.resultsBySuite[suiteName][engine] = NewEngineStats(passed, total)

	// Update totals
	if _, ok := s.totals[engine]; !ok {
		s.totals[engine] = NewEngineStats(0, 0)
	}

	totalStats := s.totals[engine]
	totalStats.Passed += passed
	totalStats.Total += total
}

func (s *TestSummary) PrintSummary() {
	fmt.Printf("\n%s\n", centerText(" Test Results Summary ", 80, "="))

	// Print header
	fmt.Printf("%-30s", "Test Suite")
	for _, engine := range types.AllEngines() {
		fmt.Printf("%15s", engine)
	}
	fmt.Printf("\n%s\n", strings.Repeat("-", 80))

	// Print results for each suite
	suiteNames := make([]string, 0, len(s.resultsBySuite))
	for name := range s.resultsBySuite {
		suiteNames = append(suiteNames, name)
	}
	sort.Strings(suiteNames)

	for _, suiteName := range suiteNames {
		fmt.Printf("%-30s", suiteName)
		for _, engine := range types.AllEngines() {
			if result, ok := s.resultsBySuite[suiteName][engine]; ok {
				fmt.Printf("%7d/%-7d", result.Passed, result.Total)
			} else {
				fmt.Printf("%15s", "N/A")
			}
		}
		fmt.Println()
	}

	// Print totals
	fmt.Printf("%s\n", strings.Repeat("-", 80))
	fmt.Printf("%-30s", "TOTAL")
	for _, engine := range types.AllEngines() {
		if total, ok := s.totals[engine]; ok {
			fmt.Printf("%7d/%-7d", total.Passed, total.Total)
		} else {
			fmt.Printf("%15s", "N/A")
		}
	}
	fmt.Println()

	// Print success rates
	fmt.Printf("%-30s", "Success Rate")
	for _, engine := range types.AllEngines() {
		if total, ok := s.totals[engine]; ok {
			fmt.Printf("%14.2f%%", float64(total.Passed)/float64(total.Total)*100)
		} else {
			fmt.Printf("%15s", "N/A")
		}
	}
	fmt.Println()
}

func (s *TestSummary) SaveJSON(filename string) error {
	type jsonResult struct {
		TestSuites map[string]map[string]*EngineStats `json:"test_suites"`
		Totals     map[string]*EngineStats            `json:"totals"`
		Timestamp  string                             `json:"timestamp"`
		GoVersion  string                             `json:"go_version"`
	}

	result := jsonResult{
		TestSuites: make(map[string]map[string]*EngineStats),
		Totals:     make(map[string]*EngineStats),
		Timestamp:  time.Now().UTC().Format(time.RFC3339),
		GoVersion:  runtime.Version(),
	}

	// Convert suite results
	for suite, results := range s.resultsBySuite {
		suiteMap := make(map[string]*EngineStats)
		for engine, stats := range results {
			suiteMap[engine.String()] = stats
		}
		result.TestSuites[suite] = suiteMap
	}

	// Convert totals
	for engine, stats := range s.totals {
		result.Totals[engine.String()] = stats
	}

	// Marshal with indentation
	data, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(filename, data, 0644)
}

func centerText(text string, width int, fill string) string {
	if len(text) >= width {
		return text
	}

	leftPad := (width - len(text)) / 2
	rightPad := width - len(text) - leftPad

	return strings.Repeat(fill, leftPad) + text + strings.Repeat(fill, rightPad)
}
